modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/db/Migration.scala

```diff
@@ -181,54 +180,62 @@ object Migration {
   def empty[F[_]: Applicative]: Migration[F] =
     Migration[F](Nil, Applicative[F].unit)
 
-  def buildBlock[F[_]: MonadThrow: DAO, I](description: Description, target: Target[I]): F[Option[Block]] =
+  def buildBlock[F[_]: MonadThrow: DAO, I](description: Description, target: Target[I]): F[List[Block]] =
     description match {
       case Description.Table(schemas) =>
-        val tableName = StringUtils.getTableName(schemas.latest)
-
-        val migrate: F[Option[Block]] = for {
-          schemaKey <- getVersion[F](tableName)
-          matches = schemas.latest.schemaKey == schemaKey
-          block <- if (matches) emptyBlock[F]
-                   else
-                     Control.getColumns[F](tableName).flatMap { (columns: List[ColumnName]) =>
-                       migrateTable(target, schemaKey, columns, schemas) match {
-                         case Left(migrationError) => MonadThrow[F].raiseError[Option[Block]](migrationError)
-                         case Right(block) => MonadThrow[F].pure(block.some)
+        val schemaKeysWithModels = foldMapMergeRedshiftSchemas(schemas)
+
+        val goodModel = schemaKeysWithModels.values.collectFirst { case model: ShredModel.GoodModel =>
+          model
+        }.get
+
+        val tableName = goodModel.tableName
+
+        val goodBlock: F[Option[Block]] =
+          for {
+            schemaKeyInTable <- getVersion[F](tableName)
+            matches = goodModel.schemaKey == schemaKeyInTable
+            block <- if (matches) emptyBlock[F]
+                     else
+                       Control.getColumns[F](tableName).flatMap[Option[Block]] { (columns: List[ColumnName]) =>
+                         migrateTable(target, schemaKeyInTable, goodModel, columns, schemas) match {
+                           case Left(migrationError) => MonadThrow[F].raiseError(migrationError)
+                           case Right(block) => Monad[F].pure(block.some)
+                         }
                        }
-                     }
-        } yield block
+          } yield block
 
-        Control.tableExists[F](tableName).ifM(migrate, Monad[F].pure(target.createTable(schemas).some))
+        val recoveryBlocks: List[Block] = schemaKeysWithModels.values
+          .collect { case model: ShredModel.RecoveryModel =>
+            model
+          }
+          .map(target.createTable)
+          .toList
+
+        val blocks: F[List[Block]] = goodBlock.map(gb => gb.fold(recoveryBlocks)(_ :: recoveryBlocks))
+
+        Control.tableExists[F](tableName).ifM(blocks, Monad[F].pure(List(target.createTable(goodModel))))
```

Comment:
Non-existing table doesn't mean that recovery is not required. It is possible to have a broken schema when we load the first event, which would require a recovery table to go with good table.